### PR TITLE
Sprite shadows - take 3

### PR DESCRIPTION
**Pull request by nashmuhandes**
New language strings:

```
DSPLYMNU_SPRITESHADOW = "Actor shadows";
OPTVAL_SPRSHADALWAYS = "Monsters and players";
```

Example files:

[SpriteShadowTestMod.zip](https://github.com/coelckers/gzdoom/files/5366098/SpriteShadowTestMod.zip)
Explicitly enables sprite shadows for Zombiemen, and force the Imps to never draw shadows

[SpriteShadowEnabler.zip](https://github.com/coelckers/gzdoom/files/5366100/SpriteShadowEnabler.zip)
An event handler-based mod that simply enables shadows for all living things

### CODE REVIEW

**Comment by nashmuhandes** [comment]
Some extra notes on how I determine whether the shadow is rendered or not:

- By default, no shadows are ever drawn.

- Modders may now add +CASTSPRITESHADOWS to explicitly set their actors to render shadows. HOWEVER (very important, read the next one)

- User may force shadows to always be drawn for monsters (and players), and MOST IMPORTANTLY, set shadows to ALWAYS be OFF.

This approach satisfies everyone - modders and users. You'll have to trust that I have everyone's interest at heart.

In other words,

1) Engine doesn't draw sprite shadows by default
2) If you are playing without mods. you are literally one click away in the option menu to enable sprite shadows for monsters
3) You can forcefully disable sprite shadows if you don't like them (or for performance reasons)

**Review by coelckers — CHANGES_REQUESTED** [review]
The checker function in its current state is a killer.
Remember, this gets called for every single sprite in every single frame, so this must be brutally optimized for performance, which means:

The function must be inlined - it must not do any preliminary work on retrieving constants before it knows it actually needs these things. Calling 'IsKindOf' for checking players is a no-go as is precalculating the various conditions for determining its course of action.

For cases like this, never trust the compiler to do the right thing - do the right thing yourself!

We already once had a case where a similarly structured function broke down the entire renderer, the end result was that I hand optimized the code instead of relying on higher level facilities. The same is needed here.

**Comment by nashmuhandes** [comment]
> We already once had a case where a similarly structured function broke down the entire renderer, the end result was that I hand optimized the code instead of relying on higher level facilities. The same is needed here.

Understood. Do you remember which function was this where you had to do the optimization, so I can learn from it?

**Comment by coelckers** [comment]
You won't be able to learn from that anymore because the code had been completely refactored out of the renderer.
It was the "VisibleToPlayer" check for determining if a dynamic light should be used, which was performed deep inside the light render loop and could cause massive slowdowns because lights can affect many surfaces, and the checks were repeated for each single one.
To solve these I moved all related code out of the actual renderer into higher level data structures where I could better control when to update the needed flags so that the renderer only needs to do a single variable check.

**Review by alexey-lysiuk — COMMENTED** [review]
It's not mandatory, but may I ask you to clean this PR from merge and especially fix-for-fix aka reverted commits?

**Comment by coelckers** [comment]
I have to concur with @alexey-lysiuk here. This needs a serious round of clean up before it can be merged.

**Comment by nashmuhandes** [comment]
Got it. The "thumbs up" response was to acknowledge the request, though I wasn't free to do it earlier. :) I'll get to it ASAP.

**Comment by madame-rachelle** [comment]
This is already being well-tested with your strife coop streams is it not?

If so, any chance you'd be willing to fix the merge conflicts here so this can be merged?

**Comment by madame-rachelle** [comment]
That is in addition to the cleanups requested, as well.

**Comment by nashmuhandes** [comment]
@madame-rachelle yes it is being live-tested on stream, as you saw. Sorry I didn't get back to this yet. Will do the cleanup after my stream (so on Sunday).

I most likely will ahead and start a new branch and redo the entire PR with cleaner and higher quality commits so that it can reviewed more conveniently at each step.

**Comment by madame-rachelle** [comment]
Cool, thank you :)

**Comment by nashmuhandes** [comment]
> This is already being well-tested with your strife coop streams is it not?
> 
> If so, any chance you'd be willing to fix the merge conflicts here so this can be merged?

Redid the entire thing, cleaner: https://github.com/coelckers/gzdoom/pull/1372
